Test Guidelines (#308)
vdaas/vald: docs/developer/style.md

# Go Style Guide in Vald

## Table of Contents

- [Introduction](#Introduction)
- [Style](#Style)
- [Test](#Test)
	- [Table-Driven-Test](#Table-Driven-Test)

## Introduction

## Style

## Test

This section will describe the rule for writing test code in Vald.

> ### Table-Driven-Test

Reviewer: ```suggestion
### The guideline of Table-Driven-Test for Vald
```

Use table-driven tests with subtests to avoid duplicating code. 

<table>

> <thead><tr><th>Bad</th><th>Good</th></tr></thead>

Reviewer: Which is better the left column is `Bad` or `Good`?:thinking:
How do you think @kevindiu ?

Reviewer 2: I think `Bad` on the left column is better because I think it is better to let user to understand the `Bad` case first and then compare with `Good`, but if we write the `Good` on the left column user may avoid the `Bad` case.

}

if want, got := "192.0.2.0", host; want != got {
    t.Errorf("host is not equals. want: %s, but got: %s", want: %s, got)
}

if want, got := "http", port; want != got {
    t.Errorf("port is not equals. want: %s, but got: %s", want: %s, got)
}
```

</td>
<td>

```go
tests := []struct {
    str string
    wantHost string
    wantPort string
} {
    ## case 1
    {
        str: "192.0.2.0:8000",
        wantHost: "192.0.2.0",
        wantPort: "8000",
    },
    ## case 2
    {
    	str: "192.0.2.0:8000",
    	wantHost: "192.0.2.0",
    	wantPort: "http",
	},
}

for _, tt := range tests {
    t.Run(tt.str, func(tt *testing.T) {
        host, port, err := net.SplitHostPort(tt.str)
        if err != nil {
            t.Errorf("error is not nil: %v", err)
        }
        if want, got := tt.wantHost, host; want != got {
            t.Errorf("host is not equals. want: %s, but got: %s", want: %s, got)
        }
        if want, got := tt.wantPort, port; want != got {
            t.Errorf("port is not equals. want: %s, but got: %s", want: %s, got)
        }
    })
}
```

</td>
</tr>
</tbody>
</table>

Table-Driven-Test makes it easy to add new test case.

> We follow the convention that the `map[string]func(*testing.T)test` is referred to as tests and each test case tt. 

Author: Thank you for your review.

> We follow the convention that the `map[string]func(*testing.T)test` is referred to as tests and each test case tt. 

Reviewer 2: ```suggestion
We define the test case table as `map[string]func(*testing.T)test`, which is referred as the test case name and the test case implementation `tt`. 
```

```go
tests := map[string]func(t *testing.T) test {
    "test case name": func(tt *testing.T) test {
        return test {

        }
    }
}
```

The following are the steps to create a Table-Driven-Test.

1. `args` structure

> If there are two or more arguments to be passed to the method, create a `args` structure. If there is one argument, do not create an `args` structure.

Reviewer 2: ```suggestion
If there are two or more arguments to be passed to the method, create a `args` structure. If there is only one argument, do not create an `args` structure.
```

```go
type args struct {
    host string
    port string
}
```

2. `field` structure

> If you create an object and test its methods, create a `field` struct if object has two or more fields to initialize. If there is one field, do not create `field` structure.

Reviewer 2: ```suggestion
If you create an object and test its methods, create a `field` struct if the object has two or more fields to initialize. If there is only one field, do not create `field` structure.
```

```go
type field struct {
    host string
    port string
}
```

3. `test` structure

`test` structure has `args` and `field` structure and `checkFunc` function. If you need one of `args` and `field` structure, create `field` and `args` structure.
The `checkFunc` function is used to check the return value of the function being tested.

```go
type test struct {
    args args
    field field
    checkFunc func(t *testing.T, err error)
}
```

> Hole example is this.

Reviewer 2: ```suggestion
Example:
```

Author: Thank you for your review

# Go Style Guide in Vald

## Table of Contents

- [Introduction](#Introduction)
- [Style](#Style)
- [Test](#Test)
	- [Table-Driven-Test](#Table-Driven-Test)

## Introduction

## Style

## Test

This section will describe the rule for writing test code in Vald.

> 

Reviewer: ```suggestion
Test guideline has the 2 important rules for continuous quality.
1. Use Table-Driven-Test
2. Test all possible patterns
```
@kevindiu Is it no problem with grammar?

Reviewer 2: I think it is better.
```suggestion
Testing guideline has 2 important rules for the coding quality and readability.
```

Reviewer: I'd like to write `Test all possible patterns`.
Where should this sentence be written?

Reviewer 2: You mean the testing coverage?

Reviewer: I mean all user possible cases

> ### The guideline of Table-Driven-Test for Vald

Reviewer: Oh, did you say `Table-Driven-Test` is better?  you said it need not to add `guideline` each section before because this document describes the guideline, right?

Author: I`m sorry :bow:
I fixed but I forgot commit....

# Go Style Guide in Vald

## Table of Contents

- [Introduction](#Introduction)
- [Style](#Style)
- [Test](#Test)
	- [Table-Driven-Test](#Table-Driven-Test)

## Introduction

## Style

## Test

Testing guideline has 2 important rules for the coding quality and readability
1. Use Table-Driven-Test
2. Keep code coverage over 85%

>    - test coverage != high testing quality, but low coverage means bad testing quality

Reviewer 2: maybe we need a sample command for user to check the coverage? 
`go test -cover ./...`
